libportal/notification.c

```diff
@@ -19,9 +19,79 @@
 
 #include "config.h"
 
+#include <fcntl.h>
+
+#include <glib/gstdio.h>
+#include <gio/gunixfdlist.h>
+
 #include "notification.h"
 #include "portal-private.h"
 
+static void
+prepare_media (GVariantBuilder *builder,
+              GVariant *serialized_media,
+              GUnixFDList **fd_list)
+{
+  g_autoptr(GVariant) value = NULL;
+  g_autoptr(GFile) file = NULL;
+  const char *key;
+  g_autofree char *path = NULL;
+  int fd, fd_in = 0;
+
+  if (g_variant_is_of_type (serialized_media, G_VARIANT_TYPE("(sv)")))
+    g_variant_get (serialized_media, "(&sv)", &key, &value);
+
+  if (key && value && strcmp (key, "file") == 0 && g_variant_is_of_type (value, G_VARIANT_TYPE_STRING))
+    file = g_file_new_for_commandline_arg (g_variant_get_string (value, NULL));
+
+  /* Pass through everything that isn't a native GFile */
+  if (!G_IS_FILE (file) || !g_file_is_native (file))
+    {
+      g_variant_builder_add_value (builder, serialized_media);
+      return;
+    }
+
+  path = g_file_get_path (file);
+  fd = g_open (path, O_RDONLY | O_CLOEXEC);
+
+  if (*fd_list == NULL)
+    *fd_list = g_unix_fd_list_new_from_array (&fd, 1);
+  else
+    fd_in = g_unix_fd_list_append (*fd_list, fd, NULL);
```

Comment:
> The file descriptor is duplicated using dup(). You keep your copy of the descriptor and the copy contained in list will be closed when list is finalized.

The fd is leaking. Use g_autofd for fd.

Reply:
You could steal this https://github.com/flatpak/xdg-desktop-portal/blob/main/src/xdp-utils.h#L178

Reply:
or bump the version but I'm not sure what the policy is here